src/mlpack/methods/cf/interpolation_policies/regression_interpolation.hpp

```diff
@@ -0,0 +1,156 @@
+
+#include <mlpack/prereqs.hpp>
+
+namespace mlpack {
+namespace cf {
+
+/**
+ * Implementation of regression-based interpolation method. Predicting a user's
+ * rating r_iu by it's neighbors' ratings can be regarded as solving linear
+ * regression of r_iu on r_iv, where v are u's neighbors.
+ *
+ * For more information, see the following paper.
+ *
+ * @code
+ * @inproceedings{bell2007improved,
+ *  title={Improved neighborhood-based collaborative filtering},
+ *  author={Bell, Robert M and Koren, Yehuda},
+ *  booktitle={KDD cup and workshop at the 13th ACM SIGKDD international
+ *      conference on knowledge discovery and data mining},
+ *  pages={7--14},
+ *  year={2007},
+ *  organization={Citeseer}
+ * }
+ * @endcode
+ */
+class RegressionInterpolation
+{
+ public:
+  /**
+   * Empty Constructor.
+   */
+  RegressionInterpolation() { }
+
+  /**
+   * Use cleanedData to perform necessary preprocessing.
+   *
+   * @param cleanedData Sparse rating matrix.
+   */
+  RegressionInterpolation(const arma::sp_mat& cleanedData)
+  {
+    const size_t userNum = cleanedData.n_cols;
+    a.set_size(userNum, userNum);
+    b.set_size(userNum, userNum);
+  }
+
+  /**
+   * The regression-based interpolation problem can be solved by a linear
+   * system of equations. This method first calculates the coefficients and
+   * contant terms for the equations and then solve the equations to get
+   * weights.
+   *
+   * @param weights Resulting interpolation weights.
+   * @param w Matrix W from decomposition.
+   * @param h Matrix H from decomposition.
+   * @param queryUser Queried user.
+   * @param neighbors Neighbors of queried user.
+   * @param similarities Similarites between query user and neighbors.
+   * @param cleanedData Sparse rating matrix.
+   */
+  void GetWeights(arma::vec& weights,
+                  const arma::mat& w,
+                  const arma::mat& h,
+                  const size_t queryUser,
+                  const arma::Col<size_t>& neighbors,
+                  const arma::vec& /* similarities*/,
+                  const arma::sp_mat& cleanedData)
+  {
+    const size_t itemNum = cleanedData.n_rows;
+    const size_t neighborNum = neighbors.size();
+
+    // Coeffcients of the linear equations used to compute weights.
+    arma::mat coeff(neighborNum, neighborNum);
+    // Constant terms of the linear equations used to compute weights.
+    arma::vec constant(neighborNum);
+
+    arma::vec userRating(cleanedData.col(queryUser));
+    const size_t support = arma::accu(userRating != 0);
+
+    // If user has no rating at all, average interpolation is used.
+    if (support == 0)
+    {
+      weights.set_size(neighbors.n_elem);
+      weights.fill(1.0 / neighbors.n_elem);
+      return;
+    }
+
+    for (size_t i = 0; i < neighborNum; i++)
+    {
+      // Calculate coefficient.
+      arma::vec iPrediction;
+      for (size_t j = i; j < neighborNum; j++)
+      {
+        if (a(neighbors(i), a(neighbors(j))) != 0)
```

Comment:
Did you mean `a(neighbors(i), neighbors(j))`?

Reply:
Oops made a strange mistake... Thanks!